src/subcommand/torrent/announce.rs

```diff
@@ -0,0 +1,135 @@
+use crate::common::*;
+
+const URI_HELP: &str = "Announce an infohash and list the response (a compact peer list).";
+
+const INPUT_HELP: &str =
+  "Generate a compact peer list from a metainfo at `INPUT`. If `INPUT` is `-`, read \
+                          metainfo from standard input.";
+
+const INPUT_FLAG: &str = "input-flag";
+
+const INPUT_POSITIONAL: &str = "<INPUT>";
+
+#[derive(StructOpt)]
+#[structopt(
+  help_message(consts::HELP_MESSAGE),
+  version_message(consts::VERSION_MESSAGE),
+  about(URI_HELP)
+)]
+pub(crate) struct Announce {
+  #[structopt(
+    name = INPUT_FLAG,
+    long = "input",
+    short = "i",
+    value_name = "INPUT",
+    empty_values(false),
+    parse(try_from_os_str = InputTarget::try_from_os_str),
+    help = INPUT_HELP,
+  )]
+  input_flag: Option<InputTarget>,
+  #[structopt(
+    name = INPUT_POSITIONAL,
+    value_name = "INPUT",
+    empty_values(false),
+    parse(try_from_os_str = InputTarget::try_from_os_str),
+    required_unless = INPUT_FLAG,
+    conflicts_with = INPUT_FLAG,
+    help = INPUT_HELP,
+  )]
+  input_positional: Option<InputTarget>,
+}
+
+impl Announce {
+  pub(crate) fn run(self, env: &mut Env, options: &Options) -> Result<(), Error> {
+    let target = xor_args(
+      "input_flag",
+      &self.input_flag,
+      "input_positional",
+      &self.input_positional,
+    )?;
+    let input = env.read(target)?;
+
+    let infohash = Infohash::from_input(&input)?;
+    let metainfo = Metainfo::from_input(&input)?;
+
+    let mut rng = rand::thread_rng();
+    let peer_id: [u8; 20] = rng.gen();
+    let mut peer_list = Vec::new();
+
+    if metainfo.trackers().peekable().peek().is_none() {
```

Comment:
To get the trackers, you can do something like this up front:

```rust
let trackers = metainfo.trackers.collect::<Vec<Result<Url, Error>>>().filter_map(|result| {
  // print error message if tracker failed to parse
  // return None if failed to parse, so it gets filtered out
}).collect::<Vec<Url>>();

if trackers.is_empty() {
  return Err(Error::MissingTrackers);
}
```

Or, actually, to keep going if any trackers are bad, we can do:

Reply:
Half your message didn't make it. How's the listing below?

```rust
let trackers = metainfo.trackers().collect::<Vec<Result<Url, Error>>>()
      .into_iter()
      .filter_map(|result| {
        match result {
          Ok(tracker) => Some(tracker),
          Err(err) => {
            if !options.quiet {
              out!(env, "{:?}", err).ok();
            }
            None
          }
        }}).collect::<Vec<Url>>();
```
It looks kinda gross to me. But maybe because I haven't drunk enough rusty kool-aid

Reply:
Yah, that's a little gross. I think a for loop is better, since it can be tricky to use `?` inside a closure:

```rust
let mut trackers = Vec::new();
for result in metainfo.trackers() {
  match result {
    Ok(tracker) => trackers.push(tracker),
    Err(err) => if !options.quiet {
      outln!("Failed to parse tracker URL: {}", err)?;
    }
  }
}
```

It's totally fine to use a for loop. Iterator chains are ideal when they're a good fit, but when you've got to return errors in the middle I tend to use a for loop.

Reply:
Updated the above to use `errln`, since it's a warning message.